bee-tangle/src/urts.rs

```diff
@@ -0,0 +1,204 @@
+        }
+    }
+
+    pub(crate) async fn update_scores<B: StorageBackend>(&mut self, tangle: &MsTangle<B>) {
+        let mut to_remove = Vec::new();
+
+        for tip in self.tips.keys() {
+            match self.tip_score::<B>(tangle, &tip).await {
+                Score::SemiLazy | Score::Lazy => {
+                    to_remove.push(*tip);
+                }
+                _ => continue,
+            }
+        }
+
+        for tip in to_remove {
+            self.tips.remove(&tip);
+            self.non_lazy_tips.remove(&tip);
+        }
+
+        debug!("Non-lazy tips {}", self.non_lazy_tips.len());
+    }
+
+    async fn tip_score<B: StorageBackend>(&self, tangle: &MsTangle<B>, message_id: &MessageId) -> Score {
+        // in case the tip was pruned by the node, consider tip as lazy
+        if !tangle.contains(message_id).await {
+            return Score::Lazy;
+        }
+
+        let smi = *tangle.get_solid_milestone_index();
+
+        // The tip pool only works with solid tips. Therefore, all tips added to the pool can be considered to solid.
+        // The solid flag will be set together with omrsi and ymrsi values. Therefore, when a message is solid, omrsi
+        // and ymrsi values are available. Therefore, unwrapping here is fine.
+        let omrsi = *tangle.omrsi(&message_id).await.unwrap().index();
+        let ymrsi = *tangle.ymrsi(&message_id).await.unwrap().index();
+
+        if (smi - ymrsi) > YMRSI_DELTA {
+            return Score::Lazy;
+        }
+
+        if (smi - omrsi) > BELOW_MAX_DEPTH {
+            return Score::Lazy;
+        }
+
+        if (smi - omrsi) > OMRSI_DELTA {
+            return Score::SemiLazy;
+        }
+
+        Score::NonLazy
+    }
+
+    pub fn two_non_lazy_tips(&self) -> Option<Vec<MessageId>> {
+        let non_lazy_tips = &self.non_lazy_tips;
+        if non_lazy_tips.is_empty() {
+            None
+        } else {
+            Some(if non_lazy_tips.len() < self.optimal_num_tips() {
+                non_lazy_tips.iter().copied().collect()
+            } else {
+                non_lazy_tips
+                    .iter()
+                    .choose_multiple(&mut rand::thread_rng(), self.optimal_num_tips())
+                    .iter()
+                    .map(|t| **t)
+                    .collect()
+            })
+        }
+    }
+
+    pub(crate) fn optimal_num_tips(&self) -> usize {
+        // TODO: hardcoded at the moment
+        4
+    }
+
+    pub(crate) fn reduce_tips(&mut self) {
+        let mut to_remove = Vec::new();
+        for (tip, metadata) in &self.tips {
+            if let Some(age) = metadata.time_first_child {
+                if age.elapsed().as_secs() > MAX_AGE_SECONDS_AFTER_FIRST_CHILD as u64 {
+                    to_remove.push(*tip);
+                }
+            }
+        }
+        for tip in to_remove {
+            self.tips.remove(&tip);
+            self.non_lazy_tips.remove(&tip);
+        }
+    }
+}
```

Comment:
Before doing something like this, I'd like to get a few things straight:

1) Mutable state is bad, and it becomes even worse when persisted. Even in a seemingly innocent case like this, it's mental overhead that wouldn't otherwise be necessary for a maintainer to have to deal with. Something like 'don't forget to clean it at the end' is an implicit contract that can't be checked by the compiler, and so gets hoisted on to the developer tasked with maintaining the thing.

2) Modern allocators are fast. *Really fast*. Especially for small allocations like this. Unless you're telling me that `to_remove` grows to multiple thousands of elements in size every time it's invoked, this is going to be small fry compared to the other hot code we're running in the node. We shouldn't over-complicate things for small fry.

3) `retain` is already stable.

Reply:
re: 3. Did I write `retain`? My bad. I mean `drain_filter`. 
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain_filter

Reply:
re: 2. The current impl is still inefficient as it will reallocate the Vec several times. 6 times on average with current speeds. But this is going to increase in the future.

Reply:
re: 1. In this case it's not that bad, because this function is the only place where this state is mutated.

Reply:
Also. This code is part of the hot-path as the tip-set changes with every incoming message, hence the need to keep the tip-pool at bay, meansing it's needs to be reduced constantly. This is exactly the type of code where you'ld want to reduce unnecessary allocations.

Reply:
```rs
    pub(crate) fn reduce_tips(&mut self) {
        let non_lazy_tips = &mut self.non_lazy_tips;
        self.tips.retain(|tip, metadata| metadata
            .time_first_child
            .filter(|age| age.elapsed().as_secs() > MAX_AGE_SECONDS_AFTER_FIRST_CHILD as u64)
            .map(|_| non_lazy_tips.remove(&tip))
            .is_none());
    }
```

Here's a better implementation that avoids mutable state

Reply:
I does...but it's not equivalent code, because it ties the `retain` to the success of the removal in `non_lazy_tips`, which the former code didn't do.

Reply:
oh no. it's fine :joy: